### PR TITLE
0012 add event database

### PR DESCRIPTION
**Pull request by realityexpander**
Sorry again for making this one so big! But its mostly boilerplate, and replication of very similar data structures in the various layers.

Android with all these layers has SO MUCH BOILERPLATE for the simplest tasks... There must be a better way than this, but I dont know what it is. I assume this layering of everything comes in handy when the applications get very large, like dozens of screens or hundreds, with all kinds of database entities interacting with each other...

Im wondering if there are code generators for a lot of this boilerplate code, like writing DTO's and Domain and Entities and the converters. Its very repetitious!

I also added a lot of "todo" comments to show my work-in-progress or future work.

* Added Photo, Attendee Domain, DTOs & Entities classes
* Added logout
* Added Local Database 
* Added table for  Agenda `Event`
* Added Fake and real `Event` `DAO`
* Hooked up UI to local database as SSOT
* Allow Creating AgendaItem Events 
* Uploading created events to server
* Added Events API interface
* Enhanced Events Repository
* Fixed tests
* Fixed Fake implementations
* Created Abstract `EventDTO` and implmented the `EventDTO`s for `Create`, `Update` and `Response`
* Added `@TypeConverters` for:
   * `ZonedDateTime`
   * lists of strings (`uuid`s, `userId`s, etc)
   * lists of  `Photo`s and `Attendees` (instead of building tables for these items and doing joins)
* Add Entity <-> Domain <-> DTO converters for `Photo` and `Attendee` and `Event`

## Question 1
Do we need to open drop-down menus upon return from process death?

## Question 2
I am planning on saving a a couple entites as JSON strings directly in the `Event` Table, as they really dont interact with other Agenda items.

Is this a good idea? It seems like a simpler approach than making a table for every entity and using joins.

### CODE REVIEW

**Review by philipplackner — APPROVED** [review]
1. Either is fine, it's just important to restore what would break the state or what would be very unexpected for the user. They probably won't even notice if a drop down is hidden when coming back.
2. It's the same what I did in my project. Attendees and photos are just a JSON serialized list using a type converter. As long as you don't need to query them (which you don't need here), it's all good and saves a ton of (maintenance) work.

File: app/src/main/java/com/realityexpander/tasky/agenda_feature/data/common/convertersDTOEntityDomain/EventConverters.kt [hunk 1]
```diff
@@ -39,44 +40,108 @@ fun EventEntity.toDomain(): AgendaItem.Event {
...
+                description = description,
+                from = from.toZonedDateTime(),
+                to = to.toZonedDateTime(),
+                remindAt = remindAt.toZonedDateTime(),
+
+                host = host,
+                isUserEventCreator = isUserEventCreator,
+                isGoing = isGoing,
+                attendees = attendees.map { it.toDomain() },
+                attendeeIds = attendees.map { it.id },      // extract the attendees Ids
```
[the remark this answers is not in the document]
Comment (realityexpander) [reply app/src/main/java/com/realityexpander/tasky/agenda_feature/data/common/convertersDTOEntityDomain/EventConverters.kt:90]: OK, I see that now. I will replace with a single property.

File: app/src/main/java/com/realityexpander/tasky/agenda_feature/data/common/convertersDTOEntityDomain/EventConverters.kt [hunk 2]
```diff
@@ -39,44 +40,108 @@ fun EventEntity.toDomain(): AgendaItem.Event {
...
+                photos = photos.map { it.toDomain() },
+                photoIds = photos.map { it.id },            // extract the photo Ids
+            )
```
[the remark this answers is not in the document]
Comment (realityexpander) [reply app/src/main/java/com/realityexpander/tasky/agenda_feature/data/common/convertersDTOEntityDomain/EventConverters.kt:93]: Will change this too.

File: app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/eventRepository/remote/eventApi/DTOs/EventDTO.kt [hunk 3]
```diff
@@ -1,35 +1,83 @@
...
+
+    // Core information for all Event DTOs
+    abstract val id: UuidStr
+    abstract val title: String
+    abstract val description: String
+    abstract val remindAt: UtcMillis
+    abstract val from: UtcMillis
+    abstract val to: UtcMillis
+
+    @Keep
```
[the remark this answers is not in the document]
Comment (realityexpander) [reply app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/eventRepository/remote/eventApi/DTOs/EventDTO.kt:20]: I will investigate the proguard solution.

File: app/src/main/java/com/realityexpander/tasky/agenda_feature/domain/Photo.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import com.realityexpander.tasky.agenda_feature.util.UrlStr
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+data class Photo(
+    val id: PhotoId,
+    val url: UrlStr,          // url of the photo on server
+
+    val uri: String? = null,  // points to a local file on device (if photo is new)
+) : Parcelable
```
[the remark this answers is not in the document]
Comment (realityexpander) [reply app/src/main/java/com/realityexpander/tasky/agenda_feature/domain/Photo.kt:14]: OK, I will implement this. I assume the `id` is identical for `Local` and `Lemote`, so they refer to the same image?

**Pull request merged**